src/main/scala/tech/sourced/gemini/Gemini.scala

```diff
@@ -122,13 +205,90 @@ class Gemini(session: SparkSession, log: Slf4jLogger, keyspace: String = Gemini.
     val uast2seqRequest = Uast2seqRequest(uast=Some(uast), docfreqThreshold=5)
     client.identifiers(idRequest)
 
-    val features = for {
-      idResponse <- client.identifiers(idRequest)
-      litResponse <- client.literals(litRequest)
-      uast2seqResponse <- client.uast2Seq(uast2seqRequest)
-    } yield idResponse.features ++ litResponse.features ++ uast2seqResponse.features
+    try {
+      val features = for {
+        idResponse <- client.identifiers(idRequest)
+        litResponse <- client.literals(litRequest)
+        uast2seqResponse <- client.uast2Seq(uast2seqRequest)
+      } yield idResponse.features ++ litResponse.features ++ uast2seqResponse.features
+
+      Await.result(features, Duration(30, SECONDS))
+    } catch {
+      case e: Exception => {
+        log.error(s"feature extractor error: ${e.toString}")
+        Iterable[Feature]()
+      }
+    }
+  }
+
+  def parseDocFreq(file: File): (Int, Map[String, Double], List[String]) = {
+    val docFreqByteArray = Files.readAllBytes(file.toPath)
+    val docFreqJson = JSON.parseFull(new String(docFreqByteArray))
```

Comment:
You might want to encapsulate this JSON parsing logic in something like:

```scala
def mustParseJSON[T: ClassTag](input: String): T = {
  JSON.parseFull(input) match {
    case Some(res: T) => res
    case Some(_) => throw new Exception("incorrect json")
    case None => throw new Exception("can't parse json")
  }
}
```

Then just:

```scala
val docFreqJson = mustParseJSON[Map[_, _]](new String(docFreqByteArray))
```

Reply:
My code would need some changes to actually be equivalent to your code with the `.asInstanceOf[Map[String, List[List[Double]]]]` stuff. But the idea would still be the same.

Reply:
I kept `mustParseJSON` as simple as you suggested because it's also temporary code until we have our own hashing.